apps/system/js/acl_manager.js

```diff
@@ -0,0 +1,283 @@
+      var app = evt.application;
+      if (this.isExternalApp(app.manifest)) {
+        debug('external app installed, attempting to connect');
+        this.attemptToConnect();
+      }
+    }.bind(this));
+  };
+
+  ACLManager.prototype.checkForExternalApps = function() {
+    var req = navigator.mozApps.mgmt.getAll();
+    req.onsuccess = function() {
+      var apps = req.result;
+      var found = apps.some(function(app) {
+        return this.isExternalApp(app.manifest);
+      }.bind(this));
+      if (found) {
+        debug('External app already installed, attempting to connect');
+        this.attemptToConnect();
+      }
+    }.bind(this);
+
+    req.onerror = function(e) {
+      debug('Unable to fetch apps ', e);
+    };
+  };
+
+  ACLManager.prototype.attemptToConnect = function() {
+    debug('Starting acl connection attempt');
+    this.pollStartedAt = Date.now();
+    this.openConnection();
+  };
+
+  ACLManager.prototype.openConnection = function() {
+    if (this.socket) {
+      debug('Cannot connect, already have socket');
+      return;
+    }
+
+    // Check for polling timeout.
+    if (Date.now() - this.pollStartedAt > POLL_TIMEOUT) {
+      debug('Poll timeout reached, giving up');
+      return;
+    }
+
+    try {
+      this.socket = navigator.mozTCPSocket.open(HOST, PORT);
+      this.socket.onopen = this.handleOpen.bind(this);
+      this.socket.onclose = this.handleClose.bind(this);
+      this.socket.ondata = this.handleData.bind(this);
+      this.socket.onerror = this.handleSocketError.bind(this);
+    } catch (e) {
+      this.handleSocketError(e);
+    }
+  };
+
+  ACLManager.prototype.handleSocketError = function(e) {
+    debug('Error connecting: ' + JSON.stringify(e));
+    this.socket = null;
+    setTimeout(this.openConnection.bind(this), RETRY_TIMEOUT);
+  };
+
+  ACLManager.prototype.handleOpen = function() {
+    debug('CONNECTION SUCCESS!');
+  };
+
+  ACLManager.prototype.handleClose = function(evt) {
+    debug('Closed connections: ' + evt.type);
+    this.socket = null;
+    setTimeout(this.openConnection.bind(this), RETRY_TIMEOUT);
+  };
+
+  ACLManager.prototype.handleData = function(evt) {
+    var data = evt.data;
+    this.buffer += data;
+    var i = this.buffer.indexOf('\n');
+    while (i !== -1) {
+      var msg = this.buffer.slice(0, i);
+      this.buffer = this.buffer.slice(i + 1);
+      this.handleMessage(msg);
+      i = this.buffer.indexOf('\n');
+    }
+  };
+
+  ACLManager.prototype.serializeMessage = function(obj) {
+    return JSON.stringify(obj) + '\n';
+  };
+
+  ACLManager.prototype.sendMessage = function(msg) {
+    if (!this.socket) {
+      debug('Cannot send message, no connection', msg);
+    }
+    this.socket.send(this.serializeMessage(msg));
+  };
+
+  ACLManager.prototype.getAppFromMessage = function(msg) {
+    var app = null;
+    if (msg.manifestURL) {
+      app = applications.getByManifestURL(msg.manifestURL);
+    }
+    if (!app && msg.origin) {
```

Comment:
We don't want these to be `else` since the above functions will return null if they can't find apps with that manifest or origin.